stdlib/GMP_jll/src/GMP_jll.jl

```diff
@@ -0,0 +1,44 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+## dummy stub for https://github.com/JuliaBinaryWrappers/GMP_jll.jl
+module GMP_jll
```

Comment:
Comparing this to https://github.com/JuliaBinaryWrappers/GAP_jll.jl/blob/main/src/GAP_jll.jl I notice a couple differences:
A very minor one is that the the real GAP_jll is a `baremodule` which supposedly helps performance a bit -- but that's irrelevant for compatibility.

More importantly, though, the JLL API implemented here seems to match the "old" one (you know, from the stone ages, i.e., 2-3 months ago). A quick check shows that it is missing at least these:
- `best_wrapper`
- `dev_jll`
- `find_artifact_dir`
- `get_libgmp_path`
- `get_libgmpxx_path`

I actually just today adapted some of my code to use `GMP_jll.find_artifact_dir()` instead of `GMP_jll.artifact_dir`, as I just learned about `find_artifact_dir` and thought this was the way forward. Now it isn't?

Perhaps it would be a good idea to use something like JLLWrappers here, too, to avoid some of the repetition and ensure consistency. Indeed, perhaps one could adapt JLLWrapper itself?

Finally, and  for me perhaps most seriously: this fake JLL will lack the GMP header files, won't it? Because those actually are quite crucial for me, as I have to compile some C code on the user machine which links against GMP_jll. Yes yes, I should put it all into JLLs, and I did it for most of the code, but there are some parts where this is rather problematic.

If there was a way to get the "real" `GMP_jll` (and similar for other JLLs in this PR), just for accessing the headers in it, that'd solve this final issue for me (in particular, it would not be necessary for me to load the "real" `GMP_jll`, just a way to get paths into it would suffice. Is something like that a possibility?

Reply:
> Now it isn't?

It is.

> If there was a way to get the "real" GMP_jll (and similar for other JLLs in this PR), just for accessing the headers in it, that'd solve this final issue for me (in particular, it would not be necessary for me to load the "real" GMP_jll, just a way to get paths into it would suffice. Is something like that a possibility?

Yes!  This is precisely why we bundle a `StdlibArtifacts.toml` file; so that "special" packages that are in the know can do these kinds of workarounds.  In your case, you'll just need to do some artifacts shenanigans if you're on v1.6, where you find the `StdlibArtifacts.toml` file, select the appropriate `GMP` artifact within it, and then you can use your headers and whatnot as usual.

> Perhaps it would be a good idea to use something like JLLWrappers here, too, to avoid some of the repetition and ensure consistency. Indeed, perhaps one could adapt JLLWrapper itself?

I would have loved to, and indeed, the first cut at this did exactly that, but there are some issues:

* First, the JLLWrappers API has evolved a lot over the past few months.  You've seen this (and helped a bit) so I knew that wholesale importing JLLWrappers into v1.6 wasn't quite a possibility yet.  IMO, better to ship no API at all then to ship a broken one that we replace two months later in v1.7.

* Second, we don't have actual artifacts getting installed on v1.6 yet.  That breaks a lot of assumptions in the JLLWrapper world.  I would like for JLLWrappers to be smart enough to deal with `USE_SYSTEM_XXX=1` configurations where you just `dlopen("libfoo.so")`, alternate artifact location configurations (e.g. a `LibFoo_jll_debug=true` is set in your `LocalPreferences.toml`) or complete path overrides (e.g. the preference `LibFoo_jll_libfoo_path=/opt/special_julia_stuff/lib/libfoo.so.1.2` is set).  But that API isn't here yet, and that's _really_ what we need in order for JLLs to be useful across all Julia configurations.

Ideally, it would all come together in one piece, but unfortunately that kind of thinking ends up with PRs that have thousands of lines changed at once, and are completely un-reviewable.  So we have to do them in small batches, and when release time comes, I can only hold it back for so long.  ;)

Unfortunately, this forces a bit of complexity onto you, but hopefully it's not _too_ too inconvenient to have some `v1.6`-specific workarounds.

Reply:
Oh, and we can add some of those API pieces in.  Adding e.g. `get_libgmp_path()` shouldn't be too hard, and `get_artifact_dir()` can just return `dirname(Sys.BINDIR)` or something.

Reply:
Thanks for clarifying. Great to hear about `StdlibArtifacts.toml`, that's a relief :-)

Reply:
Just to give a little higher level context: 1.6 is going to be a somewhat awkward middle state for stdlib JLLs which will be resolved in 1.7 by making stdlib JLLs really be fully normal JLLs, but it's a big change, so not all of it made it into the 1.6 release. I'm really looking forward to consistently using real JLLs everywhere.